src/bridge/cockpitpipechannel.c

```diff
@@ -483,3 +510,43 @@ cockpit_pipe_channel_open (CockpitTransport *transport,
   json_object_unref (options);
   return channel;
 }
+
+static void
+internal_fd_free (gpointer data)
+{
+  gint fd = GPOINTER_TO_INT (data);
+
+  close (fd);
+}
+
+const gchar *
+cockpit_pipe_channel_add_internal_fd (gint fd)
+{
+  /* We are not multi-threaded. Also don't make this look like normal fd numbers */
+  static guint64 unique = 911111;
+  gboolean inserted;
+
+  gchar *id;
+
+  if (!internal_fds)
+    internal_fds = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, internal_fd_free);
+
+  id = g_strdup_printf ("file-descriptor-%" G_GUINT64_FORMAT, unique++);
```

Comment:
This name is a bit misleading. For folks debugging this, they'll think that the file descriptor N is being opened. Worth choosing another name like 'internal-stream-XXX' or something like that? If you don't think so, I'm fine with keeping it as it is. Just consider it.